app/utils/validation.ts

```diff
@@ -109,23 +113,59 @@ export const ifNotField =
     context[field] ? ([true] as const) : validator(value, context);
 
 export function createValidator(
-  fieldValidators: { [field: string]: ReturnType<Validator>[] },
-  rawValidator?: (input) => ValidatorResult
+  fieldValidators: {
+    [field: string]: (ReturnType<Validator> | ReturnType<AsyncValidator>)[];
+  },
+  rawValidator?: (input) => ValidatorResult,
+  makeAsync = false
 ) {
-  return function validate(input) {
-    const rawValidatorErrors: ValidatorResult = rawValidator?.(input) || {};
-    const fieldValidatorErrors = Object.keys(fieldValidators).reduce(
-      (errors: ValidatorResult, field) => {
-        const fieldErrors =
-          fieldValidators[field]
-            .map((validator) => validator(input[field], input))
-            .filter(([isValid]) => !isValid)
-            .map(([, message]) => message || 'not valid') || 0;
-        if (fieldErrors.length) errors[field] = fieldErrors;
-        return errors;
-      },
-      {}
-    );
-    return merge(fieldValidatorErrors, rawValidatorErrors);
-  };
+  return makeAsync
+    ? async function validate(input) {
```

Comment:
can we just do `if (makeAsync)` instead? Turneries are just more difficult to read when they get this big